### PR TITLE
feat(query): add plan cache

### LINKED ISSUE #16294: bug: COPY INTO <location> does not support tables that need quoting
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

v1.2.618-nightly

### What's Wrong?

Trying to COPY INTO gcs from table with dashes in name gives error. Works fine for table with no dashes

### How to Reproduce?

```sql
CREATE table t1 (c1 int null);
CREATE table "t-2" (c1 int null);

select * from default.t1;

COPY INTO 'gcs://projects/tables/t1'
CONNECTION = (
	CREDENTIAL = '<snip>'
)
FROM default.t1
FILE_FORMAT = (TYPE = PARQUET);

select * from default."t-2";

COPY INTO 'gcs://projects/tables/t-2'
CONNECTION = (
	CREDENTIAL = '<snip>'
)
FROM default."t-2"
FILE_FORMAT = (TYPE = PARQUET);
```
```
SQL:1:32
  |
1 | SELECT * FROM default.default.t-2
  |                                ^ unexpected `-`, expecting `AT`, `CHANGES`, `WITH`, <Ident>, <LiteralString>, `IDENTIFIER`, `AS`, `PIVOT`, `UNPIVOT`, `SAMPLE`, `ROW`, `BLOCK`, `(`, `LiteralAtString`, `LATERAL`, `NATURAL`, `INNER`, `LEFT`, `SEMI`, `RIGHT`, `ANTI`, `FULL`, `CROSS`, `JOIN`, `ON`, `USING`, `,`, `WHERE`, `GROUP`, `HAVING`, `WINDOW`, `QUALIFY`, `UNION`, `EXCEPT`, `INTERSECT`, `FROM`, `SELECT`, `VALUES`, `ORDER`, `LIMIT`, `OFFSET`, `IGNORE_RESULT`, `FORMAT`, or `;`

} cause: null
```
### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

Add plan cache, queries will reuse the plan from cache if source plan is not changed (snapshot is not changed).

This cache is similar to Oracle's Global Plan Cache, but it's stored as LRU cache in memory.

Also, fix the name resolution issue of #16294 
fixes #16294 

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [x] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [x] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16333)

### CODE REVIEW

File: src/query/sql/src/planner/planner_cache.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+use sha2::Digest;
+use sha2::Sha256;
+
+use crate::normalize_identifier;
+use crate::plans::Plan;
+use crate::NameResolutionContext;
+use crate::PlanExtras;
+use crate::Planner;
+
+type PlanCache = LruCache<String, (Plan, PlanExtras)>;
```
Comment (zhang2014) [inline src/query/sql/src/planner/planner_cache.rs:38]:
```suggestion
type PlanCache = InMemoryLruCache<String, (Plan, PlanExtras)>;
```

**Review by xudong963 — COMMENTED** [review]
Do we need unit tests to verify 
1. get plan from cache
2.  if the table structure changed, can't get plan from cache?

**Comment by sundy-li** [comment]
> Do we need unit tests to verify
get plan from cache
if the table structure changed, can't get plan from cache?

The setting is default enabled, so logictest covered both issues.

**Comment by BohuTANG** [comment]
> > Do we need unit tests to verify
> > get plan from cache
> > if the table structure changed, can't get plan from cache?
> 
> The setting is default enabled, so logictest covered both issues.

It would be better to add some new logic tests for this PR to ensure it remains unbreakable in the future, as we’re not sure if the existing tests cover this PR or if those tests might change later on.

**Review by xudong963 — APPROVED** [review]

**Pull request merged**